### PR TITLE
[CCP-185] Changes for displaying alternative ids in all rest calls

### PR DESCRIPTION
**Pull request by mkarthik90**

### CODE REVIEW

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/AlternativeIdService.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+import edu.asu.conceptpower.app.core.ConceptTypesService;
+import edu.asu.conceptpower.app.core.ConceptTypesService.ConceptTypes;
+import edu.asu.conceptpower.core.ConceptEntry;
+
+@Service
+public class AlternativeIdService {
+
+    @Autowired
+    private ConceptTypesService conceptTypesService;
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/AlternativeIdService.java]: I just realize that, but this class needs an interface and be moved into the correct package.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/AlternativeIdService.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import edu.asu.conceptpower.app.core.ConceptTypesService;
+import edu.asu.conceptpower.app.core.ConceptTypesService.ConceptTypes;
+import edu.asu.conceptpower.core.ConceptEntry;
+
+@Service
+public class AlternativeIdService {
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/AlternativeIdService.java]: javadoc

Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/AlternativeIdService.java]: interface missing

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/AlternativeIdService.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            entry.getAlternativeIds().add(entry.getId());
+        }
+        // Specific Wordnet id is added irrespective of what is queried for
+        if (entry.getWordnetId() != null) {
+            String[] wordNetIds = entry.getWordnetId().split(",");
+            for (String wordNetId : wordNetIds) {
+                entry.getAlternativeIds().add(wordNetId);
+            }
+        }
+    }
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/AlternativeIdService.java]: alternative ids should include all ids, shouldn't it? So ConceptId, WordnetId, and if general wordnet id was queried for that as well. Right now, you're not adding the concept id, do you? And what about merged ids?

**Comment by jdamerow** [comment]
Also needs test cases

**Comment by mkarthik90** [comment]
@jdamerow  Addressed review comments.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/AlternativeIdService.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    private IConceptTypesService conceptTypesService;
+
+    /**
+     * This method adds the alternative ids provided concept entry and the id
+     * with which the entry is queried is passed as parameter.
+     * 
+     * Alternative ids comprises of wordnet ids, local concept id, merged id and
+     * generic wordnet id(if concept entry is queried with generic wordnet id
+     * ??)
+     */
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/AlternativeIdService.java]: javadoc is incomplete. missing: parameter description, return type descriptoin

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/AlternativeIdService.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+ */
+@Service
+public class AlternativeIdService implements IAlternativeIdService {
+
+    @Autowired
+    private IConceptTypesService conceptTypesService;
+
+    /**
+     * This method adds the alternative ids provided concept entry and the id
+     * with which the entry is queried is passed as parameter.
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/AlternativeIdService.java]: This sentence isn't really understandable. Do you mean: "This method adds all alternative ids to the given concept entry including the id that was used to search for a concept (passed as method parameter)." ?

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/AlternativeIdService.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+            String[] mergedIds = entry.getMergedIds().split(",");
+            for (String mergedId : mergedIds) {
+                entry.getAlternativeIds().add(mergedId.trim());
+            }
+        }
+    }
+
+    /**
+     * This method in turn calls the addAlternativeIds(String queriedId,
+     * ConceptEntry entry) to add the alternative id
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/AlternativeIdService.java]: not a good javadoc description. missing parameter and return type description and a description of what this method really does (not how it's implemented).

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/AlternativeIdService.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    public void addAlternativeIds(Collection<ConceptEntry> conceptEntries) {
+        for (ConceptEntry entry : conceptEntries) {
+            addAlternativeIds(entry.getId(), entry);
+        }
+    }
+
+    /**
+     * This method in turn calls the addAlternativeIds(String queriedId,
+     * ConceptEntry entry) to add the alternative id
+     */
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/AlternativeIdService.java]: same here

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/AlternativeIdService.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    public void addAlternativeIds(ConceptEntry[] conceptEntries) {
+        for (ConceptEntry entry : conceptEntries) {
+            addAlternativeIds(entry.getId(), entry);
+        }
+    }
+}
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/AlternativeIdService.java]: also, the javadoc should probably be on the interface and inherited by the class.

**Comment by mkarthik90** [comment]
Added javadocs.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/IAlternativeIdService.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+import java.util.Collection;
+
+import edu.asu.conceptpower.core.ConceptEntry;
+
+public interface IAlternativeIdService {
+
+    /**
+     * This method adds the alternative ids to the concept entry which is passed
+     * as a method parameter.
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/IAlternativeIdService.java]: Is the queriedId parameter added to the alternative ids as well?

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/IConceptTypesService.java [hunk 10]
```diff
@@ -0,0 +1,8 @@
+package edu.asu.conceptpower.app.core;
+
+import edu.asu.conceptpower.app.core.impl.ConceptTypesService.ConceptTypes;
+
+public interface IConceptTypesService {
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/IConceptTypesService.java]: javadoc

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/AlternativeIdService.java [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    @Autowired
+    private IConceptTypesService conceptTypesService;
+
+    /**
+     * This method adds all alternative ids to the given concept entry including
+     * the id that was used to search for a concept (passed as method
+     * parameter).
+     * 
+     * If the id passed as a method parameter is null then null value is added
+     * to alternative ids
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/AlternativeIdService.java]: why?

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/IConceptTypesService.java [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+
+    /**
+     * This method returns the concept type based on the id parameter.
+     * 
+     * @param id
+     * @return ConceptTypes
+     */
+    public ConceptTypes getConceptTypeByConceptId(String id);
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/IConceptTypesService.java]: The return type is defined in the class implementing this interface => Circular dependency (class implements interface, but interface needs type defined in class) == bad.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/IConceptTypesService.java [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+import edu.asu.conceptpower.app.core.impl.ConceptTypesService.ConceptTypes;
+
+public interface IConceptTypesService {
+
+    /**
+     * This method returns the concept type based on the id parameter.
+     * 
+     * @param id
+     * @return ConceptTypes
+     */
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/IConceptTypesService.java]: javadoc should be a bit more specific. what are the rules that decide what id is what type?

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/ConceptTypesService.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+@Service
+public class ConceptTypesService implements IConceptTypesService {
+    
+    /**
+     * This enum contains the list of valid concept types.
+     * 
+     * @author karthikeyanmohan
+     *
+     */
+    public enum ConceptTypes {
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/ConceptTypesService.java]: Name of enum should be singular (one instance of the enum represents one type).

Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/ConceptTypesService.java]: Also, it's not really a concept type, but the type of id used to refer to a concept.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/ConceptTypesService.java [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+        GENERIC_WORDNET_CONCEPT, SPECIFIC_WORDNET_CONCEPT, LOCAL_CONCEPT
+    }
+    
+    /**
+     * This method returns the concept types based on the id. If id is null then
+     * null value is returned by the method.
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/ConceptTypesService.java]: "concept types" should a) be singular, and b) it's not really a concept type (we have those and they are something else), but the type of id used to refer to a concept.

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/app/core/impl/AlternativeIdServiceTest.java [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+    @Before
+    public void setUp() {
+        MockitoAnnotations.initMocks(this);
+    }
+
+    @Test
+    public void test_addAlternativeIds_successQueriedParameterIsNull() {
+        String queriedId = null;
+        Mockito.when(conceptTypesService.getConceptTypeByConceptId(queriedId)).thenReturn(null);
+        alternativeIdService.addAlternativeIds(queriedId, null);
```
Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/app/core/impl/AlternativeIdServiceTest.java]: add a comment saying why you don't assert anything

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/IConceptTypesService.java (hunk 13 above)
Comment (mkarthik90) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/IConceptTypesService.java]: I think it's implementation specific to specify how the concept type is determined, so I have added that in classes level doc.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/IAlternativeIdService.java (hunk 9 above)
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/IAlternativeIdService.java]: this is still not answers. also, what are the alternative ids?

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/IConceptTypesService.java [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+public interface IConceptTypesService {
+    
+    /**
+     * This enum contains the list of valid concept types.
+     * 
+     * @author karthikeyanmohan
+     */
+    public enum TypeOfId {
+        GENERIC_WORDNET_CONCEPT, SPECIFIC_WORDNET_CONCEPT, LOCAL_CONCEPT
+    }
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/IConceptTypesService.java]: should be "IdType"

Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/IConceptTypesService.java]: and then rename the enums accordingly

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/AlternativeIdService.java [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+    /**
+     * This method adds all alternative ids to the given concept entry including
+     * the id that was used to search for a concept (passed as method
+     * parameter).
+     * 
+     * If id or concept entry passed to the method is null, then method just
+     * returns and no changes are made.
+     * 
+     * The queriedId parameter will be added to the alternative id list if the
+     * queriedId is of Generic wordnet concept type.
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/AlternativeIdService.java]: paragraph 1 and 3 are generally true (should be true for any implementation of the interface), so they should be on the interface. However, paragraph 1 and 3 are also contradicting:
Para 1: "adds all alternative ids to the given concept entry including the id that was..."
Para 3: "will be added to the alternative id list if the queriedId is of Generic wordnet concept type."
Which one is it?

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/ConceptTypesService.java [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+     * 
+     * Concept Type is determined based on the type of id.
+     * 
+     * If id contains ??, then that concept is a generic wordnet concept.
+     * 
+     * If id starts with WID, then that concept is a specifc wordnet concept.
+     * 
+     * If both of the above condition is not met, then the concept is local CCP
+     * concept.
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/core/impl/ConceptTypesService.java]: this is true for any implementation of the interface (or at least the different types of ids). so the interface should at least list the different types.

**Comment by jdamerow** [comment]
almost there. Javadoc can be hard ;)

**Comment by mkarthik90** [comment]
@jdamerow Addressed review comments.

**Comment by jdamerow** [comment]
just a few small things

**Comment by mkarthik90** [comment]
@jdamerow  Addressed review comments.

**Comment by jdamerow** [comment]
ok to test

**Comment by jdamerow** [comment]
run integration tests please

**Pull request merged**